### PR TITLE
never_patterns: lower a never arm to an unreachable terminator

### PR DESCRIPTION
**Pull request by Nadrieril**
This introduces `ExprKind::Unreachable` (in hir and thir), which is lowered to an unreachable terminator in mir. For the non-trivial cases (mir lowering and type checking), I modeled it after the case of a match expression without arms, and the result looks sensible.

I then use this to lower never arms. This doesn't change the final optimized mir because the `UninhabitedEnumBranching` pass would detect the empty variant anyway, but this way is cleaner.

r? @compiler-errors

### CODE REVIEW

File: tests/mir-opt/match/never_patterns.rs [hunk 1]
```diff
@@ -0,0 +1,17 @@
+// skip-filecheck
+#![feature(never_patterns)]
+#![allow(incomplete_features)]
+
+enum Void {}
+
+// EMIT_MIR never_patterns.opt1.UninhabitedEnumBranching.before.mir
```
Comment (cjgillot) [inline tests/mir-opt/match/never_patterns.rs:7]: Could this be a MIR building test, outputting the just-built mir ?

Comment (Nadrieril) [reply tests/mir-opt/match/never_patterns.rs:7]: Yes! How do I do that?

**Comment by Nadrieril** [comment]
I'm a bit disappointed at the `match x { ! }` diagnostic change but I don't know how to avoid it

File: tests/mir-opt/match/never_patterns.rs (hunk 1 above)
Comment (Nadrieril) [reply tests/mir-opt/match/never_patterns.rs:7]: Is it just `EMIT_MIR never_patterns.opt1.built.after.mir`?

File: compiler/rustc_ast/src/util/parser.rs [hunk 2]
```diff
@@ -287,6 +287,7 @@ pub enum ExprPrecedence {
     Struct,
     Gen,
     Await,
+    Unreachable,
```
Comment (compiler-errors) [inline compiler/rustc_ast/src/util/parser.rs:290]: Can you just give `ExprKind::Unreachable` a dummy precedence? Like, just give it the same precedence as `Lit` or something.

File: compiler/rustc_ast_lowering/src/expr.rs [hunk 3]
```diff
@@ -570,19 +570,9 @@ impl<'hir> LoweringContext<'_, 'hir> {
                 self.dcx().emit_err(NeverPatternWithGuard { span: g.span });
             }
 
-            // We add a fake `loop {}` arm body so that it typecks to `!`.
-            // FIXME(never_patterns): Desugar into a call to `unreachable_unchecked`.
```
Comment (compiler-errors) [inline compiler/rustc_ast_lowering/src/expr.rs:574]: Your PR description doesn't explain the challenges of just lowering a call to, for example, the `unreachable` intrinsic. I feel like that's far less invasive than adding a new HIR expr kind.

**Pull request closed without merging**